core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java

```diff
@@ -271,25 +278,58 @@ public void process(StreamDecoderInput in, StreamDecoderOutput<WebSocketFrame> o
                     if (frameOpcode == WebSocketFrameType.TEXT.opcode()) {
                         decodedFrame = WebSocketFrame.ofPooledText(payloadBuffer, finalFragment);
-                        out.add(decodedFrame);
                     } else if (frameOpcode == WebSocketFrameType.BINARY.opcode()) {
                         decodedFrame = WebSocketFrame.ofPooledBinary(payloadBuffer, finalFragment);
-                        out.add(decodedFrame);
-                    } else if (frameOpcode == WebSocketFrameType.CONTINUATION.opcode()) {
-                        decodedFrame = WebSocketFrame.ofPooledContinuation(payloadBuffer, finalFragment);
-                        out.add(decodedFrame);
                     } else {
-                        throw protocolViolation(WebSocketCloseStatus.INVALID_MESSAGE_TYPE,
-                                                "Cannot decode a web socket frame with opcode: " + frameOpcode);
+                        assert frameOpcode == WebSocketFrameType.CONTINUATION.opcode();
+                        decodedFrame = WebSocketFrame.ofPooledContinuation(payloadBuffer, finalFragment);
                     }
                     logger.trace("{} is decoded.", decodedFrame);
+
                     if (finalFragment) {
                         fragmentedFramesCount = 0;
+                        aggregatingFramesLength = 0;
+                        if (aggregatingFrames.isEmpty()) {
+                            out.add(decodedFrame);
+                        } else {
+                            aggregatingFrames.add(decodedFrame);
+                            final ByteBuf[] byteBufs = aggregatingFrames.stream()
+                                                                        .map(Bytes::byteBuf)
+                                                                        .toArray(ByteBuf[]::new);
+                            if (aggregatingFrames.get(0).type() == WebSocketFrameType.TEXT) {
+                                out.add(WebSocketFrame.ofPooledText(Unpooled.wrappedBuffer(byteBufs), true));
+                            } else {
+                                out.add(WebSocketFrame.ofPooledBinary(Unpooled.wrappedBuffer(byteBufs), true));
+                            }
+                            aggregatingFrames.clear();
+                        }
                     } else {
                         fragmentedFramesCount++;
+                        if (aggregateContinuation) {
+                            aggregatingFramesLength += framePayloadLength;
+                            if (aggregatingFramesLength > maxFramePayloadLength) {
+                                throw protocolViolation(
+                                        WebSocketCloseStatus.MESSAGE_TOO_BIG,
+                                        // The message must not exceed 125 length:
+                                        // https://datatracker.ietf.org/doc/html/rfc6455/#section-5.5
```

Comment:
nit; this comment probably isn't related since it refers to control frames only

```suggestion
```

Reply:
The message is set to the reason phrase of the close frame and the close frame is one of control frames so I left the comment here. 😉 
However, I've changed to logic to truncate the message if it exceeds the limit in `WebSocketUtil` so I've moved this comment to there. PTAL. 😉